### PR TITLE
Upgrade to .NET 8.0 and eradicate ☠ Mono ☠

### PR DESCRIPTION
**Pull request by michaeldgg2**
### Changes

Apart from targeting .NET 8.0 this PR does other things too:
- removes all Mono-related stuff
- updates README.md
- updates `.editorconfig` after migrating to .NET 8.0 and removing Mono
- fixes some warnings related to upgrade to .NET 8.0
- removes workarounds for Mono in code

### TODO
- fix new warnings after changes to `.editorconfig`
- someone with macOS should finish and test macOS stuff (I made separate commit for that to separate it from other changes)
- script for dedicated server (?) for FreeBSD still uses Mono (someone with FreeBSD should migrate it to the new .NET)

### .NET 8.0 on older Windows versions

Surprisingly, not only .NET 8.0 runtime works on Windows 8.1, but also the SDK. More importantly OpenRA compiled targeting .NET 8.0 works on Windows 8.1 as well (this runs in a VM so ignore the FPS):

openra_net80_win81

### CODE REVIEW

**Comment by RoosterDragon** [comment]
The current dev appetite is for the next release to still use net6+mono, so this will need to wait until after that has occured at a minimum.

I would recommend grepping the repo for `net6`, `net 6` and `mono` for additional areas the will need an update. Between documentation, packaging and other various scripts and code config I would expect a complete switchover to net 8 and dropping of mono to affect dozens of files.

File: Makefile [hunk 1]
```diff
@@ -50,13 +47,12 @@ gameinstalldir ?= $(libdir)/openra
...
 MSBUILD = msbuild -verbosity:m -nologo
 DOTNET = dotnet
-MONO = mono
 RM = rm
 RM_R = $(RM) -r
 RM_F = $(RM) -f
 RM_RF = $(RM) -rf
 
-RUNTIME ?= net6
+RUNTIME ?= net8
```
Comment (Mailaender) [inline Makefile]: If the only runtime is `net8` then maybe get rid of this all together.

**Comment by Mailaender** [comment]
It saves into `~/Dokumente/.config/openra/` instead of `~/.config/openra/` on Linux.

**Review by Mailaender — CHANGES_REQUESTED** [review]
It saves into `~/Documents/.config/openra/` instead of `~/.config/openra/` on Linux. https://learn.microsoft.com/dotnet/core/compatibility/core-libraries/8.0/getfolderpath-unix

File: .editorconfig [hunk 2]
```diff
@@ -820,7 +820,7 @@ dotnet_diagnostic.CA1843.severity = warning
 # Provide memory-based overrides of async methods when subclassing 'Stream'.
 dotnet_diagnostic.CA1844.severity = warning
 
-# Use span-based 'string.Concat'. (Not available on mono)
+# Use span-based 'string.Concat'.
 dotnet_diagnostic.CA1845.severity = none
```
Comment (RoosterDragon) [inline .editorconfig]: Let's set this to a warning.

File: .gitignore [hunk 3]
```diff
@@ -25,10 +25,6 @@ IP2LOCATION-LITE-DB1.IPV6.BIN.ZIP
 \#*
 .*.sw?
 
-# Monodevelop
-*.pidb
-*.userprefs
-
```
Comment (RoosterDragon) [inline .gitignore:31]: Changes related to the Monodevelop IDE can stay.

File: INSTALL.md [hunk 4]
```diff
@@ -78,6 +76,6 @@ Type `sudo make install` for system-wide installation. Run `sudo make install-li
 macOS
 =====
 
-[.NET 6](https://dotnet.microsoft.com/download/dotnet/6.0) or [Mono](https://www.mono-project.com/download/stable/#download-mac) (version 6.12 or later) is required to compile OpenRA. We recommend using .NET 6 unless you are running a very old version of macOS (10.9 through 10.14).
+[.NET 8](https://dotnet.microsoft.com/download/dotnet/6.0) is required to compile OpenRA.
```
Comment (RoosterDragon) [inline INSTALL.md]: URL not updated

File: OpenRA.Game/FieldLoader.cs [hunk 5]
```diff
@@ -26,7 +25,6 @@ public static class FieldLoader
 	{
 		const char SplitComma = ',';
 
-		[Serializable]
```
Comment (RoosterDragon) [inline OpenRA.Game/FieldLoader.cs:29]: There's a couple of other `[Serializable]` attributes against Exception types that we may as well delete as well.

File: OpenRA.Game/Support/AssemblyLoader.cs [hunk 6]
```diff
@@ -9,9 +9,6 @@
  */
 #endregion
 
-// Not used/usable on Mono. Only used for Dotnet Core.
-// Based on https://github.com/natemcmaster/DotNetCorePlugins and used under the terms of the Apache 2.0 license
-#if NET5_0_OR_GREATER
```
Comment (RoosterDragon) [inline OpenRA.Game/Support/AssemblyLoader.cs:14]: There are some other `#if NET5_0_OR_GREATER` and `#if !NET5_0_OR_GREATER` statements across the codebase that can be simplified as well.

File: OpenRA.sln [hunk 7]
```diff
@@ -72,11 +72,4 @@ Global
 	GlobalSection(ExtensibilityGlobals) = postSolution
 		SolutionGuid = {9E2FAACA-DADB-4EEB-9898-2B0269F45E69}
 	EndGlobalSection
-	GlobalSection(MonoDevelopProperties) = preSolution
```
Comment (RoosterDragon) [inline OpenRA.sln:75]: Changes related to the Monodevelop IDE can stay.

File: configure-system-libraries.sh [hunk 8]
```diff
@@ -16,39 +16,22 @@ patch_config()
...
-				fi
-			done
-		done
-	else
-		# .NET 6 does not support .config files, so we must use symlinks instead
-		# Exit early if the symlink already exists
-		if [ -L "bin/${REPLACE}" ]; then
-			return 0
-		fi
+	# .NET 6 does not support .config files, so we must use symlinks instead
```
Comment (RoosterDragon) [inline configure-system-libraries.sh:19]: `.NET 8` - assuming the comment still holds in 8.

**Review by RoosterDragon — COMMENTED** [review]
Still a few hits for `net 6` in these files:

- .github\workflows\documentation.yml
- .github\workflows\packaging.yml
- packaging\functions.sh
- packaging\macos\apphost.c
- packaging\windows\buildpackage.sh

And `mono` in these files:
- OpenRA.Game\Exts.cs
- packaging\freebsd\openra

Extra `System` or `Microsoft` nuget packages that align with the runtime version should be bumped from 6 to 8.

File: .gitignore (hunk 3 above)
Comment (pchote) [reply .gitignore:31]: Monodevelop is EOL, might as well remove?

File: OpenRA.sln (hunk 7 above)
Comment (Mailaender) [reply OpenRA.sln:75]: https://github.com/mono/monodevelop is dead too.

Comment (pchote) [reply OpenRA.sln:75]: Likewise here.

File: configure-system-libraries.sh (hunk 8 above)
Comment (pchote) [reply configure-system-libraries.sh:19]: With mono gone we don't have .config files at all, so this whole file needs to be updated (comments at top, function names, etc) to reflect .NET.

File: packaging/functions.sh [hunk 9]
```diff
@@ -11,7 +11,7 @@
 #   SRC_PATH: Path to the root OpenRA directory
 #   DEST_PATH: Path to the root of the install destination (will be created if necessary)
 #   TARGETPLATFORM: Platform type (win-x86, win-x64, osx-x64, osx-arm64, linux-x64, linux-arm64, unix-generic)
-#   RUNTIME: Runtime type (net6, mono)
+#   RUNTIME: Runtime type (net8)
```
Comment (pchote) [inline packaging/functions.sh]: The RUNTIME arg should be removed completely?

**Comment by PunkPun** [comment]
needs a rebase for support folder to be correct on Mac and Linux

**Comment by pchote** [comment]
This would also be a good time to get rid of `configure-system-libraries.sh` and follow the normal .NET behaviours for native libraries.

File: packaging/macos/buildpackage.sh [hunk 10]
```diff
@@ -127,8 +125,6 @@ modify_plist "{MINIMUM_SYSTEM_VERSION}" "10.11" "${TEMPLATE_DIR}/Contents/Info.p
 # Compile universal (x86_64 + arm64) arch-specific apphosts
 clang apphost.c -o "${TEMPLATE_DIR}/Contents/MacOS/apphost-x86_64" -framework AppKit -target x86_64-apple-macos10.15
 clang apphost.c -o "${TEMPLATE_DIR}/Contents/MacOS/apphost-arm64" -framework AppKit -target arm64-apple-macos10.15
-clang apphost-mono.c -o "${TEMPLATE_DIR}/Contents/MacOS/apphost-mono" -framework AppKit -target x86_64-apple-macos10.11
```
Comment (pchote) [inline packaging/macos/buildpackage.sh:130]: The `MINIMUM_SYSTEM_VERSION` line just above will also need to be changed to the minimum version that .NET8 runs on (officially macOS 12, but I see references on the dotnet repo to it working on 10.15 and 11).

File: packaging/macos/launcher.m [hunk 11]
```diff
@@ -203,31 +159,22 @@ - (void)launchGameWithArgs: (NSArray *)gameArgs
...
-			dllPath = [exePath stringByAppendingPathComponent: @"arm64/OpenRA.dll"];
-		}
-		else
-		{
-			launchPath = [exePath stringByAppendingPathComponent: @"apphost-x86_64"];
-			hostPath = [exePath stringByAppendingPathComponent: @"x86_64/libhostfxr.dylib"];;
-			dllPath = [exePath stringByAppendingPathComponent: @"x86_64/OpenRA.dll"];
-		}
-	}
+    cpu_type_t type;
```
Comment (abcdefg30) [inline packaging/macos/launcher.m]: Spaces -> Tabs in this file please.

File: packaging/functions.sh [hunk 12]
```diff
@@ -11,7 +11,7 @@
 #   SRC_PATH: Path to the root OpenRA directory
 #   DEST_PATH: Path to the root of the install destination (will be created if necessary)
 #   TARGETPLATFORM: Platform type (win-x86, win-x64, osx-x64, osx-arm64, linux-x64, linux-arm64, unix-generic)
```
Comment (pchote) [inline packaging/functions.sh]: IMO this would be a good time to also drop the `win-x86` build.

**Comment by dnqbob** [comment]
umm, is Mono dead or something bad happened?

**Comment by JovialFeline** [comment]
> umm, is Mono dead or something bad happened?

Something like that. https://www.mono-project.com/

**Comment by dnqbob** [comment]
thanks! @JovialFeline

File: packaging/macos/buildpackage.sh (hunk 10 above)
Comment (pchote) [reply packaging/macos/buildpackage.sh:130]: Confirmed that the net8 test build runs on 10.15, so `10.11` &rarr; `10.15` above.

**Comment by michaeldgg2** [comment]
Somebody should take a look at FreeBSD start script too.

**Comment by omnivagant** [comment]
.NET 6 is out of support since yesterday https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core

File: packaging/macos/buildpackage.sh [hunk 13]
```diff
@@ -212,7 +208,7 @@ for MOD in "Red Alert" "Tiberian Dawn"; do
 done
 
 for MOD in "Red Alert" "Tiberian Dawn" "Dune 2000"; do
-	for p in "arm64" "mono"; do
+	for p in "arm64"; do
```
Comment (PunkPun) [inline packaging/macos/buildpackage.sh:211]: cool loop

**Review by PunkPun — COMMENTED** [review]
Looks like Exts has a function that references mono ResizeArray. It's unused so the whole function could be just removed.

Could you also rebase for the new SDL?

**Pull request closed without merging**